### PR TITLE
Implement case reallocation

### PR DESCRIPTION
**Pull request by C-gyorfi**
### Context
This is the first pass of implementing case reallocation to another force, capturing the logic from the legacy Bichard application. Also includes a couple of minor bug fixes and changes to the amendment logic.

### Changes
- Add a new service function, `reallocateCourtCaseToForce` that:
  - Updates the `ForceOwner` structure on the AHO
  - Add manual force code to AHO
  - Sets the `orgForPoliceFilter` to the new force
  - Adds notes for the amendments
  - Adds notes for the reallocation
  - Doing all this with the transaction so if any of the steps fail the operation rolls back
- [Bug fix: All cases displayed when visible forces fields is empty](https://github.com/ministryofjustice/bichard7-next-ui/pull/224/commits/066e19f7083bc959a3be81d183cab755c7fadda1)
- [Bug fix: Show cases based on visible courts, similar to old UI](https://github.com/ministryofjustice/bichard7-next-ui/pull/224/commits/5dbf134d82e276c7fbee61d239b06322c363b2bf)
- Changes to amendment logic:
  - Update the complete ForceOwner structure when updating force owner, similar to the logic in the legacy app
  - Generate notes for all amended fields
  - Only allow amendments when the case is visible to the user
  - Prevent amending cases that are locked by someone else
- Added a placeholder UI for reallocation and a cypress test, these will change once we have a design!

### Next steps:
- Reallocation generates a trigger (TRPR0028) in some cases. There is a [separate ticket for this](https://dsdmoj.atlassian.net/browse/BICAWS-1846) 
- Add audit log messages when a case is reallocated(Old bichard pushes messages to GENERAL_EVENT_QUEUE). There is a ticket to [add audit logs to the new UI](https://dsdmoj.atlassian.net/browse/BICAWS-2777)
- Finalise the UI and update cypress tests for reallocation

### CODE REVIEW

**Review by agnieszkazbikowska-madetech — CHANGES_REQUESTED** [review]
It looks really well! Well done! I left a few comments but very small ones.

File: cypress/e2e/court-cases/caseDetails.cy.ts [hunk 1]
```diff
@@ -357,6 +357,13 @@ describe("Case details", () => {
       .eq(0)
       .find("td")
       .last()
+      .should("have.text", "Bichard02: Portal Action: Update Applied. Element: nextHearingDate. New Value: 2024-09-26")
```
[the remark this answers is not in the document]
Comment (C-gyorfi) [reply cypress/e2e/court-cases/caseDetails.cy.ts:360]: I think id leave this just to make sure the notes displayed in the right order

File: src/utils/amendments/amendForceOwner/amendForceOwner.test.ts [hunk 2]
```diff
@@ -8,7 +8,7 @@ describe("amend force owner", () => {
 
   beforeEach(() => {
     aho = createDummyAho() as AnnotatedHearingOutcome
-    testString = "RANDOM_FORCE_OWNER_TEST_STRING"
+    testString = "04CA"
   })
 
   it("amend force owner when force owner property exist", () => {
```
[the remark this answers is not in the document]
Comment (C-gyorfi) [reply src/utils/amendments/amendForceOwner/amendForceOwner.test.ts]: Updated in dcd1d86571a85c7cba8f9225e384a2b908f8d31d

File: src/utils/amendments/getSystemNotes.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import Note from "services/entities/Note"
+import User from "services/entities/User"
+import { Amendments } from "types/Amendments"
+import { formatDisplayedDate } from "utils/formattedDate"
+
+const formatValue = (value: string | number | boolean | Date): string =>
```
[the remark this answers is not in the document]
Comment (C-gyorfi) [reply src/utils/amendments/getSystemNotes.ts]: Updated [8d5599f](https://github.com/ministryofjustice/bichard7-next-ui/pull/224/commits/8d5599f533e47448f8c4052fc618897c38391915)

File: src/utils/amendments/getSystemNotes.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+  for (const [key, value] of Object.entries(amendments)) {
+    if (key === "noUpdatesResubmit") {
+      continue
+    }
+
+    if (Array.isArray(value)) {
+      value.forEach((field) => {
+        notes.push({
+          noteText: `${portalActionText} Element: ${key}. New Value: ${formatValue(field.updatedValue)}`,
```
[the remark this answers is not in the document]
Comment (C-gyorfi) [reply src/utils/amendments/getSystemNotes.ts]: Updated in [3555d2b](https://github.com/ministryofjustice/bichard7-next-ui/pull/224/commits/3555d2b2683707a5baa940b045209aafa1c5998f)

File: test/services/reallocateCourtCaseToForce.integration.test.ts [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+      await insertCourtCasesWithFields([courtCase])
+
+      const user = {
+        username: "Dummy User",
+        visibleForces: [oldForceCode],
+        canLockExceptions: true,
+        canLockTriggers: true
+      } as User
+
+      ;(insertNotes as jest.Mock).mockImplementationOnce(() => new Error(`Error while creating notes`))
```
[the remark this answers is not in the document]
Comment (C-gyorfi) [reply test/services/reallocateCourtCaseToForce.integration.test.ts:177]: I think it is a convention for these one-liners, set by the linter

File: test/services/amendCourtCase.integration.test.ts [hunk 6]
```diff
@@ -94,30 +103,128 @@ describe("amend court case", () => {
...
+    expect(retrievedCase?.notes[1].noteText).toEqual(
+      `${userName}: Portal Action: Update Applied. Element: courtOffenceSequenceNumber. New Value: 3333`
+    )
+    expect(retrievedCase?.notes[2].userId).toEqual("System")
+    expect(retrievedCase?.notes[2].noteText).toEqual(
+      `${userName}: Portal Action: Update Applied. Element: courtOffenceSequenceNumber. New Value: 1111`
+    )
+  })
+
+  it("should not generate system note when its a no update resubmit amendment", async () => {
```
[the remark this answers is not in the document]
Comment (C-gyorfi) [reply test/services/amendCourtCase.integration.test.ts]: Fixed [61a06c7](https://github.com/ministryofjustice/bichard7-next-ui/pull/224/commits/61a06c7fe6ecd03c28b948dd4bab6496e7228195)

File: test/services/amendCourtCase.integration.test.ts [hunk 7]
```diff
@@ -151,24 +259,44 @@ describe("amend court case", () => {
...
+      phase: 1,
+      orgForPoliceFilter: orgCode
+    })
+
+    const result = await amendCourtCase(dataSource, { noUpdatesResubmit: true }, inputCourtCase.errorId, user)
+
+    expect(result).toEqual(Error(`Failed to get court case`))
+  })
+
+  it("return an error when the case is not visible to the user", async () => {
```
[the remark this answers is not in the document]
Comment (C-gyorfi) [reply test/services/amendCourtCase.integration.test.ts]: Fixed [61a06c7](https://github.com/ministryofjustice/bichard7-next-ui/pull/224/commits/61a06c7fe6ecd03c28b948dd4bab6496e7228195)

File: test/services/reallocateCourtCaseToForce.integration.test.ts [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    jest.resetAllMocks()
+    jest.clearAllMocks()
+    ;(insertNotes as jest.Mock).mockImplementation(jest.requireActual("services/insertNotes").default)
+  })
+
+  afterAll(async () => {
+    await dataSource.destroy()
+  })
+
+  describe("when user can see the case", () => {
```
[the remark this answers is not in the document]
Comment (C-gyorfi) [reply test/services/reallocateCourtCaseToForce.integration.test.ts]: Fixed [61a06c7](https://github.com/ministryofjustice/bichard7-next-ui/pull/224/commits/61a06c7fe6ecd03c28b948dd4bab6496e7228195)

**Review by agnieszkazbikowska-madetech — APPROVED** [review]

**Pull request merged**